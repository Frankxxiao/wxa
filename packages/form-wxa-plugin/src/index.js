/**
 * form plugin for wxa.js
 * 
 * <input class="" wxa-validate type="text" data-rule="required" data-as="" data-name="x.y" data-scope="scope">
 * 
 * <button disabled="{{$errorBag.scope.valid x.y}}"></button>
 * 
 * <bankcard > 
 * 
 */
import * as Rules from './rules/index'

class Validator {
    
    constructor() {
        this.defaultOptions = {
            locale: 'en',
            dictionary: null
        }
        this.options = Object.assign({}, this.defaultOptions, Validator.options);
    }

    static set(name, options) {
        this[name] = options;
    }

    static findName(name) {
        // return name.
    }

    validate(name, value, rule) {
        let validator = Rules[rule];

    }

    // findName()
}

export default (options)=>{
    console.log('===')
    Validator.set('options', options);

    return (vm, type)=>{
        if(~['Page', 'Component'].indexOf(type)) {

            vm.$validator = new Validator()
            vm.$validate = function(e) {
                let {currentTarget: {dataset: {name, as, rule, scope}}, detail: {value=''}} = e;

                if(name == null) {
                    console.warn('you must setup data-name for wxa-validator');
                    return ;
                }

                value = value.replace(/^\s*|\s*$/g, '');
                this.setData({  
                    [name]: value
                });
                this.$validator.validate(name, value, rule);
            }
            vm.$validator.validate;
        }
    };
};
