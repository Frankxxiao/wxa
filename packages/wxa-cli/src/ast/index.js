import traverse from '@babel/traverse';
import * as t from '@babel/types';
import path from 'path';
import PathParser from '../helpers/pathParser';
import {getDistPath, isFile, readFile, isDir} from '../utils';
import findRoot from 'find-root';
import resolveAlias from '../helpers/alias';
import DependencyResolver from '../helpers/dependencyResolver';
import debugPKG from 'debug';

let debug = debugPKG('WXA:ASTManager');

export default class ASTManager {
    constructor(resolve, meta) {
        debug('resolve %o ', resolve);
        debug('meta %o ', meta);

        this.resolve = resolve;
        this.meta = meta;

        this.modulesPath = path.join(meta.current, 'node_modules', path.sep);
        this.npmPath = path.join(meta.current, meta.dist, 'npm', path.sep);
    }

    parse(mdl) {
        debug('start');
        if (mdl.ast == null) return [];

        let self = this;
        let libs = [];
        traverse(mdl.ast, {
            CallExpression(path) {
                // commandJS module
                if (
                    path.node.callee &&
                    path.node.callee.name === 'require' &&
                    path.node.arguments.length
                ) {
                    let dep = path.node.arguments[0].value;

                    let dr = new DependencyResolver(self.meta.current, self.resolve, self.meta);

                    let meta = dr.resolveDep(dep, mdl, {needFindExt: true});
                    let resolved = dr.getResolved(meta.lib, meta.source, meta.target, mdl);

                    libs.push({
                        src: meta.source,
                        pret: meta.pret,
                        $$meta: meta,
                        from: {
                            $$ASTPath: path,
                            $$category: 'ast',
                            resolved,
                        },
                    });

                    // path.replaceWithSourceString(`require(${resolved})`);
                }
            },
            ImportDeclaration(path) {
                // es module
                if (
                    path.node.source &&
                    path.node.source.value
                ) {
                    let dep = path.node.source.value;

                    let dr = new DependencyResolver(self.meta.current, self.resolve, self.meta);

                    let meta = dr.resolveDep(dep, mdl, {needFindExt: true});
                    let resolved = dr.getResolved(meta.lib, meta.source, meta.target, mdl);

                    libs.push({
                        src: meta.source,
                        pret: meta.pret,
                        $$meta: meta,
                        from: {
                            $$ASTPath: path,
                            $$category: 'ast',
                            resolved,
                        },
                    });

                    // replace source node
                    // path.get('source').replaceWith(t.stringLiteral(resolved));
                }
            },
        });

        debug('dependencies libs %O', libs);
        return libs;
    }
}
