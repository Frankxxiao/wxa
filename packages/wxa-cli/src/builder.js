import 'babel-polyfill';

import {getConfig, getFiles, readFile, isFile} from './utils';
import path from 'path';
import crypto from 'crypto';
import chokidar from 'chokidar';
import schedule from './schedule';
import logger from './helpers/logger';
import compilerLoader from './loader';
import compiler from './compilers/index';
import debugPKG from 'debug';
import {green} from 'chalk';
import defaultPret from './const/defaultPret';
import Optimizer from './optimizer';
import Generator from './generator';
import Compiler from './compilers/index';
import {AsyncParallelHook} from 'tapable';

let debug = debugPKG('WXA:Builder');

class Builder {
    constructor(src, dist, ext) {
        this.current = process.cwd();
        this.src = src || 'src';
        this.dist = dist || 'dist';
        this.ext = ext || '.wxa';
        this.isWatching = false;
        this.isWatchReady = false;
        this.queue = {};

        this.hook = {
            done: new AsyncParallelHook(['dependencies']),
        };
    }
    async init(cmd) {
        // 加载编译器
        const configs = getConfig();
        schedule.set('wxaConfigs', configs || {});

        // Todo: custome package manager, such as yarn.
        // npmManager.setup(category)

        // mount loader
        return compilerLoader
        .mount(configs.use, cmd);
    }
    filterModule(arr) {
        return arr.reduce((ret, dep)=>{
            if (
                !/src\/_wxa/.test(dep.src)
            ) {
                ret.push(dep.src);
            }


            return ret;
        }, []);
    }
    watch(cmd) {
        if (this.isWatching) return;
        this.isWatching = true;

        // set mode
        schedule.set('mode', 'watch');

        let files = this.filterModule(schedule.$indexOfModule);
        console.log(files);

        let watch = chokidar.watch(files, {
            awaitWriteFinish: {
                stabilityThreshold: 300,
                pollInterval: 100,
            },
        })
        .on('all', async (event, filepath)=>{
            if (this.isWatchReady && ['change', 'add'].indexOf(event)>-1 && !this.queue[filepath]) {
                console.log(event, filepath);
                let mdl = schedule.$indexOfModule.find((module)=>module.src===filepath);
                let isChange = true;
                // module with code;
                if (mdl.code) {
                    let content = readFile(mdl.src);
                    let md5 = crypto.createHash('md5').update(content).digest('hex');

                    mdl.code = content;
                    isChange = mdl.hash !== md5;
                }

                if (isChange) {
                    schedule.$depPending.push(mdl);
                    let changedDeps = await schedule.$doDPA();

                    let generator = new Generator(schedule.wxaConfigs.resolve, schedule.meta, schedule.wxaConfigs);
                    let generateTasks = changedDeps.map((mdl)=>{
                        return generator.do(mdl);
                    });

                    await Promise.all(generateTasks);

                    logger.message('Compile', '编译完成', true);
                } else {
                    logger.message('Complete', '文件无变化', true);
                }
                // cmd.file = path.join('..', filepath);
                // // schedule
                // logger.message(event, filepath, true);
                // this.queue[filepath] = event;
                // cmd.category = event;
                // this.build(cmd);
                // setTimeout(()=>this.queue[filepath]=false, 500);
            }
        })
        .on('ready', (event, filepath)=>{
            this.isWatchReady = true;
            logger.message('Watch', '准备完毕，开始监听文件', true);
        });
    }
    async build(cmd) {
        schedule.set('src', this.src);
        schedule.set('dist', this.dist);
        schedule.set('cmdOptions', cmd);

        logger.infoNow('Compile', 'AT: '+new Date().toLocaleString(), void(0));

        // find app.js、 app.wxa first
        let appJSON = path.join(this.current, this.src, 'app.json');
        let appJS = path.join(this.current, this.src, 'app.js');
        let wxaJSON = path.join(this.current, this.src, 'app'+this.ext);
        let isWXA = false;
        if (isFile(appJSON)) {
            isWXA = false;
        } else if (!isFile(appJSON) && isFile(wxaJSON)) {
            isWXA = true;
        } else {
            logger.errorNow('不存在app.json或app.wxa文件!');
        }

        try {
            // initial loader.
            await this.init(cmd);

            // read entry file.
            let entryMdl = {
                src: wxaJSON,
                type: 'wxa',
                category: 'App',
                pret: defaultPret,
            };
            let p;
            if (isWXA) {
                p = async ()=>{
                    let compiler = new Compiler(schedule.wxaConfigs.resolve, schedule.meta);
                    let {wxa} = await compiler.$parse(void(0), void(0), wxaJSON, 'wxa', entryMdl);

                    return wxa;
                };
            } else {
                p = ()=>Promise.resolve({
                    script: {
                        code: readFile(appJS),
                    },
                    config: {
                        code: readFile(appJSON),
                    },
                });
            }

            let ret = await p();
            let rst = ret.rst || ret;
            entryMdl.rst = rst;

            let appConfigs = JSON.parse(rst.config.code);
            // mount to schedule.
            schedule.set('appConfigs', appConfigs);
            schedule.set('$pageArray', [entryMdl]);

            // do dependencies analysis.
            await schedule.doDPA();
            debug('schedule dependencies Tree is %O', schedule.$indexOfModule);

            // module optimize, dependencies merge, minor.
            let optimizer = new Optimizer(schedule.wxaConfigs.resolve, schedule.meta);
            let optimizeTasks = schedule.$indexOfModule.map((dep)=>{
                return optimizer.do(dep);
            });

            await Promise.all(optimizeTasks);

            // module dest, dependencies copy,
            let generator = new Generator(schedule.wxaConfigs.resolve, schedule.meta, schedule.wxaConfigs);
            let generateTasks = schedule.$indexOfModule.map((mdl)=>{
                return generator.do(mdl);
            });

            await Promise.all(generateTasks);

            // done.
            await this.hook.done.promise(schedule.$indexOfModule);

            console.log(cmd);
            if (cmd.watch) this.watch();
        } catch (e) {
            console.error(e);
        }
    }
}


export default Builder;
