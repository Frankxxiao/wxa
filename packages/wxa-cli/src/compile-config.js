import {getDistPath, writeFile, readFile, error, isFile, warn} from './utils';
import {info} from './utils';
import path from 'path';
import {AsyncSeriesHook} from 'tapable';
import PathParser from './helpers/pathParser';
import logger from './helpers/logger';
import schedule from './schedule';

class CConfig {
    constructor(src, dist) {
        this.current = process.cwd();
        this.src = src || 'src';
        this.dist = dist || 'dist';
        this.type = 'json';
        this.$sourceType = 'json';
        this.code = '';
        this.modulesPath = path.join(this.current, 'node_modules');
        this.localVisualPath = path.join(this.current, 'local', path.sep);
        this.npmPath = path.join(this.current, dist, 'npm', path.sep);
        this.localPath = path.join(this.current, dist, 'local', path.sep);
        this.hooks = {
            optimizeAssets: new AsyncSeriesHook(['opath', 'compilation']),
        };
    }

    resolveComponents(code, opath, options={}) {
        if (!code.usingComponents) return code;

        let coms = code.usingComponents;
        // logger.infoNow('resolve ', opath);
        Object.keys(coms).forEach((key)=>{
            let com = code.usingComponents[key];
            let pret = new PathParser().parse(com);
            // logger.infoNow('com', com);
            // logger.infoNow('pret', pret);
            // logger.infoNow('options ', options);

            let source;
            let target;
            // relative path
            if (pret.isRelative) {
                source = path.join(opath.dir, com);
                if (options.type === 'npm') {
                    target = path.join(this.npmPath, path.relative(this.modulesPath, source));
                } else {
                    target = getDistPath(source, void(0), this.src, this.dist);
                    // logger.infoNow('target', target);
                }
            } else if (pret.isNodeModule) {
                source = path.join(this.modulesPath, com);
                target = path.join(this.npmPath, com);
            } else {
                // other plugin do not parse.
                coms[key] = com;
                return;
            }

            // copy npm or local components, generate new path;
            // checkout component extensions;
            let type = pret.isRelative ? void(0) : pret.isNodeModule ? 'npm' : void(0);
            let comOptions = {...options};
            comOptions.type = type || options.type;
            let extOfCom = ['.wxml', '.wxss', '.js', '.json'];
            // wxa com support
            let wxaFile = source+'.wxa';

            if (isFile(wxaFile)) {
                schedule.addTask(path.parse(wxaFile), void(0), comOptions );
            } else {
                // console.log(com)
                extOfCom.forEach((ext)=>{
                    let uri = source+ext;

                    // logger.infoNow('uri', uri);
                    if (isFile(uri)) {
                        // while js file use script compiler.
                        // nested components.
                        schedule.addTask(path.parse(uri), void(0), comOptions);
                        // file is exits
                        let target = path.join(this.npmPath, com+ext);
                        logger.info(`write com`, path.relative(this.current, target));
                        // writeFile(target, readFile(uri));
                    } else if (ext === '.json') {
                        logger.warn(com+'组件不存在json配置文件');
                    }
                });
            }
            // let target = path.join(pret.isNodeModule ? this.npmPath : this.localPath, com);
            let dist = getDistPath(opath, 'json', this.src, this.dist);
            // logger.infoNow('dist', dist);

            let resolved = './'+path.relative(path.parse(dist).dir, target).replace(/\\/g, '/');
            // logger.infoNow('resolved ', resolved);

            coms[key] = resolved;
        });
        code.usingComponents = coms;
        return code;
    }

    compile(content, opath, options) {
        // logger.infoNow('config to Compile', opath);

        if (content == null) {
            content = readFile(path.join(opath.dir, opath.base));
            if (content == null) throw new Error('打开文件失败 ', path.join(opath.dir, opath.base));
        }

        try {
            content = JSON.parse(content);
            // 编译组件
            content = this.resolveComponents(content, opath, options);
        } catch (e) {
            console.error(e);
            logger.errorNow(`ERROR IN: ${opath.dir+path.sep+opath.base}`, e);
            return Promise.reject(e);
        }

        this.code = JSON.stringify(content, void(0), 4);
        return this.hooks.optimizeAssets.promise(opath, this).then((err)=>{
            if (err) return Promise.reject(err);
            let target = getDistPath(opath, 'json', this.src, this.dist);
            logger.info('Config', path.relative(this.current, target));
            writeFile(target, this.code);
        }).catch((e)=>{
            logger.errorNow('Error In: '+path.join(opath.dir, opath.base), e);
        });
    }
}

export default CConfig;
