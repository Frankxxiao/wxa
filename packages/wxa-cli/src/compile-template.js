import {getDistPath, writeFile, readFile, info, error} from './utils';
import {amazingCache} from './utils';
import {AsyncSeriesHook} from 'tapable';
import compilerLoader from './loader';
import path from 'path';
import logger from './helpers/logger';
import schedule from './schedule';
import PathParser from './helpers/pathParser';

export default class CTemplate {
    constructor(src, dist, ext, options) {
        this.current = process.cwd();
        this.src = src;
        this.dist = dist;
        this.ext = ext;

        this.$sourceType = 'wxml';
        this.options = options || {};

        this.hooks = {
            optimizeAssets: new AsyncSeriesHook(['opath', 'compilation']),
        };
    }
    resolveDeps(content, opath) {
        return content.replace(
            // /([\.][\t\n\s]*)?require\([']([\w\d_\-\.\/@]+)['"]\)/ig,
            /(?:\/\*[\s\S]*?\*\/|(?:[^\\:]|^)\/\/.*)|(\.)?src\s*=\s*['"]([\w\d_\-\.\/@]+)['"]/igm,

            (match, point, lib)=>{
                // a.require()
                if (point) return match;
                // ignore comment
                if (point == null && lib == null) return match;

                let pret = new PathParser().parse(lib);

                if (pret.isRelative || pret.isAbsolute) {
                    let source = pret.isRelative ? path.join(opath.dir, lib) : path.join(this.src, lib);

                    schedule.addTask(path.parse(source));
                }

                return match;
        });
        // return content;
    }

    compile(rst, opath) {
        if (typeof rst === 'string') {
            let filepath = path.join(opath.dir, opath.base);
            rst = {
                type: rst,
                src: filepath,
                code: readFile(filepath) || '',
            };
        }

        let compiler = compilerLoader.get(rst.type);
        return amazingCache({
            source: rst.code,
            options: {configs: {}},
            transform: function(source, options) {
                return compiler.parse(source, options);
            },
        }, this.options.cache).then((succ)=>{
            let code;
            if (typeof succ === 'string') {
                code = succ;
            } else {
                code = succ.code;
            }
            // console.log('编译前', code);
            this.code = this.resolveDeps(code, path.parse(rst.src));
            this.$sourceFrom = rst.type;
            // console.log(this)
            return this.hooks.optimizeAssets.promise(opath, this).then((err)=>{
                if (err) Promise.reject(err);

                // console.log('编译后', this.code);
                let target = getDistPath(path.parse(rst.src), 'wxml', this.src, this.dist);
                logger.info('write', path.relative(this.current, target));
                writeFile(target, this.code);
            });
        }).catch((e)=>{
            console.error(e);
            logger.errorNow('Error In: '+path.join(opath.dir, opath.base), e);
        });
    }
}
