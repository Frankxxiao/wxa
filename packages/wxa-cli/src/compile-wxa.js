import path from 'path';
import {readFile, error, warn, isFile, isEmpty} from './utils';
import {DOMParser} from 'xmldom';
import Coder from './helpers/coder';
import schedule from './schedule';
import logger from './helpers/logger';

class CompileWxa {
    constructor(src, dist, ext, options) {
        this.current = process.cwd();
        this.src = src;
        this.dist = dist;
        this.ext = ext;
        this.options = options;
    }
    compile(opath, configs) {
        return this.$compile(opath, configs);
    }
    $compile(opath, configs) {
        let wxa = this.resolveWxa(opath);
        // console.log(wxa);
        if (!wxa) return Promise.reject();

        let filepath = path.join(opath.dir, opath.base);
        if (filepath === path.join(this.current, this.src, 'app' + this.ext)) {
            // let appConfig = JSON.parse('"'+wxa.config+'"');
            console.log(JSON.parse(wxa.config.code));
            delete wxa.template;
        }

        schedule.addTask(opath, wxa, configs);

        return Promise.resolve();
    }

    resolveWxa(xml, opath) {
        let filepath;

        if (typeof xml === 'object' && xml.dir) {
            opath = xml;
            filepath = path.join(opath.dir, opath.base);
        } else {
            opath = path.parse(xml);
            filepath = xml;
        }

        let content = readFile(filepath);

        if (content == null) {
            error('打开文件失败:'+filepath);
            return null;
        }

        if (content == '') return null;

        let coder = new Coder();

        let encodeXml = (content, start, endId, isTemplate=false)=>{
            while (content[start++] !== '>') {};

            return isTemplate ?
                coder.encodeTemplate(content, start, content.indexOf(endId)-1)
                :
                coder.encode(content, start, content.indexOf(endId)-1);
        };

        let startScript = content.indexOf('<script') + 7;
        content = encodeXml(content, startScript, '</script>');

        let startConfig = content.indexOf('<config') + 7;
        content = encodeXml(content, startConfig, '</config>');

        if (content.indexOf('<template') > -1) {
            let startTemplate = content.indexOf('<template') + 9;
            content = encodeXml(content, startTemplate, '</template>', true);
        }

        xml = this.parserXml(opath).parseFromString(content);

        let rst = {
            style: [],
            template: {
                code: '',
                src: '',
                type: 'wxml',
            },
            script: {
                code: '',
                src: '',
                type: 'js',
            },
            config: {
                code: '',
                src: '',
                type: 'config',
            },
        };

        // console.log(xml);

        Array.prototype.slice.call(xml.childNodes || []).forEach((child)=>{
            const nodeName = child.nodeName;
            if (nodeName === 'style' || nodeName === 'template' || nodeName === 'script' || nodeName === 'config') {
                let rstTypeObject;

                if (nodeName === 'style') {
                    rstTypeObject = {code: ''};
                    rst[nodeName].push(rstTypeObject);
                } else {
                    rstTypeObject = rst[nodeName];
                }
                rstTypeObject.src = child.getAttribute('src');
                rstTypeObject.type = child.getAttribute('lang') || child.getAttribute('type');

                if (isEmpty(rstTypeObject.type)) {
                    let map = {
                        style: 'scss',
                        template: 'wxml',
                        script: 'js',
                        config: 'json',
                    };
                    rstTypeObject.type = map[nodeName];
                }

                if (rstTypeObject.src) rstTypeObject.src = path.resolve(opath.dir, rstTypeObject.src);

                if (rstTypeObject.src && isFile(rstTypeObject.src)) {
                    const code = readFile(rstTypeObject.src);
                    if (code == null) throw new Error('打开文件失败:', rstTypeObject.src);
                    else rstTypeObject.code += code;
                } else {
                    Array.prototype.slice.call(child.childNodes||[]).forEach((code)=>{
                        if (nodeName !== 'template') {
                            rstTypeObject.code += coder.decode(code.toString());
                        } else {
                            rstTypeObject.code += coder.decodeTemplate(code.toString());
                        }
                    });
                }

                if (!rstTypeObject.src) rstTypeObject.src = path.join(opath.dir, opath.base);
            }
        });

        return rst;
    }

    parserXml(opath) {
        return new DOMParser({
            errorHandler: {
                warn(x) {
                    logger.errorNow('XML警告:'+(opath.dir+path.sep+opath.base));
                    logger.warnNow(x);
                },
                error(x) {
                    logger.errorNow('XML错误:'+(opath.dir+path.sep+opath.base));
                    logger.errorNow(x);
                },
            },
        });
    }
}

export default CompileWxa;
