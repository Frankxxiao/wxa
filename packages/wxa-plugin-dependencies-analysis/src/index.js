import { Stats } from './Stats';
import { start } from './server';
import { writeFileSync } from 'fs';

export class DependenciesAnalysisPlugin {
  constructor(options = {}) {
    this.DAPserver = null;
    this.options = options;
    this.statsHandler = new Stats();
  }

  apply(compiler = {}) {
    try {
      if (!compiler.hooks || 
        (!compiler.hooks.done && !compiler.hooks.finishRebuildModule)
      ) return;

      const DAP_NAME = 'DependenciesAnalysisPlugin';

      compiler.hooks.done.tapAsync(DAP_NAME, (compilation, next) => {
        try{
          const statsJson = this.statsHandler.get(compilation.$indexOfModule);
          this.DAPserver = start(statsJson, this.options);
          next();
        } catch(e) {
          next(e);
        }
      });
  
      compiler.hooks.finishRebuildModule.tapAsync(DAP_NAME, (compilation, changedModule, next) => {
        try {
          const statsJson = this.statsHandler.get(compilation.$indexOfModule);
          this.DAPserver.updateData(statsJson);

          next();
        } catch (e) {
          next(e);
        }
        
      });
    } catch(err) {
      console.error(err);
    }
  }
}