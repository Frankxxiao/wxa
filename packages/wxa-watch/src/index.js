/**
 * watch plugin for wxa.js
 * 
 * Usage: 
 *  class A {
        data = {
            a: {
                b: 1
            }
        }

        watch = {
            'a.b'(){
            }
        }
    }
 */

import WatchJS from 'melanke-watchjs';

export default ()=>{
    return (vm, type)=>{
        if (type === 'Page') {
            let onload = vm.onLoad || function() {};
            let onUnload = vm.onUnload || function() {};

            let watcher = [];
            let isInit = false;

            vm.onLoad = function(...args) {
                if (this.watch && !isInit && this.data) {
                    isInit = true;

                    Object.keys(this.watch).forEach((key)=>{
                        let keyArr = key.split(/[.\[\]]/).filter((x)=>x!=null&&x!=='');

                        let tar = keyArr[keyArr.length-1];

                        let source = keyArr.slice(0, -1).reduce((prev, key)=>{
                            if(typeof prev === 'object') {
                                prev = prev[key];
                                return prev;
                            } else {
                                return void(0);
                            }
                        }, this.data);

                        if(source) {
                            watcher.push([source, tar, (prop, action, newValue, oldValue)=>{
                                this.watch[key].apply(this, [newValue, oldValue]);
                            }]);
                        }
                    });
                }
                watcher.forEach((subscriber)=>{
                    WatchJS.watch(subscriber[0], [subscriber[1]], subscriber[2]);
                });

                onload.apply(this, args);
            };

            vm.onUnload = function(...args) {
                watcher.forEach((subscriber)=>{
                    WatchJS.unwatch(subscriber[0], [subscriber[1]], subscriber[2]);
                });

                onUnload.apply(this, args);
            };
        }
    };
};
