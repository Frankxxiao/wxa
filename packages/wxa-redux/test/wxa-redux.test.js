import {wxaRedux, combineReducers} from '../src/index'

getCurrentPages = function() {
    return [{route: 'pages/index/index'}];
};

let og = getApp;

let todo = function(state=[], action) {
    switch(action.type) {
        case 'add' : {
            state = [...state, action.payload]
            return state;
        }
        default : return state;
    }
}

let todoDel = function(state=[], action) {
    switch(action.type) {
        case 'del' : {
            state = [...state, action.payload]
            return state;
        }
        default : return state;
    }
}

let setData = function(data) {
    this.data = {
        ...this.data,
        ...data
    }
}

test('mount redux store and middlewares', ()=>{
    let reduxFn = wxaRedux({
        reducers: combineReducers({todo})
    });

    let mw = jest.fn();
    let reduxFnMW = wxaRedux({
        reducers: combineReducers({todo}),
        middlewares: [function(store){
            return function(){
                mw();
            }
        }]
    });

    let vm = {}, vm2 = {};

    reduxFn(vm, 'App');

    reduxFnMW(vm2, 'App');

    expect(vm.$store).not.toBeNull();
    expect(mw.mock.calls.length).toBe(1);
});

test('throw error while type undefined', ()=>{
    let reduxFn = wxaRedux({}, 'Hello');

    let page = {};

    expect(()=>{
        reduxFn(page)
    }).toThrowErrorMatchingSnapshot();
});

describe('Page', ()=>{
    let reduxFnApp = wxaRedux({
        reducers: combineReducers({
            todo,
            todoDel
        })
    });
    let app = {}; reduxFnApp(app, 'App');
    
    let reduxFnPage = wxaRedux({});

    global.getApp = ()=>app;

    test('subscribe redux', ()=>{

        let page = {
            app,
            setData
        }; 
    
        reduxFnPage(page, 'Page');
    
        expect(page.$unsubscribe).toBeFalsy();
        page.onLoad();
        expect(page.$unsubscribe).toBeDefined();
    });

    test('call with default function', ()=>{
        const onLoad = jest.fn();
        const onShow = jest.fn();
        const onHide = jest.fn();
        const onUnload = jest.fn();

        let page = {
            app,
            setData,
            onLoad, onShow, onHide, onUnload
        };

        reduxFnPage(page, 'Page');

        expect(onLoad.mock.calls.length).toBe(0);
        page.onLoad({scene: 110});
        expect(onLoad.mock.calls.length).toBe(1);
        expect(onLoad).toHaveBeenCalledWith({scene: 110});
        page.onLoad();
        page.onLoad();
        expect(onLoad.mock.calls.length).toBe(3);
        
        expect(onShow.mock.calls.length).toBe(0);
        page.onShow();
        expect(onShow.mock.calls.length).toBe(1);
        expect(page.$$isCurrentPage).toBe(true);
        
        expect(onHide.mock.calls.length).toBe(0);
        page.onHide();
        expect(onHide.mock.calls.length).toBe(1);
        expect(page.$$isCurrentPage).toBe(false);
        
        expect(onUnload.mock.calls.length).toBe(0);
        page.onUnload();
        expect(onUnload.mock.calls.length).toBe(1);
        expect(page.$unsubscribe).toBeNull();

    })

    test('update data', ()=>{
         let page = {
             app,
             mapState: {
                 todoList: (state)=>state.todo
             },
             setData
         };

         reduxFnPage(page, 'Page');

         page.onLoad();

         page.$store.dispatch({type: 'add', payload: 'test+1'});
         expect(page.data).toBeFalsy();
         expect(page.$store.getState().todo.length).toBe(1);
         
         page.onShow();
         page.$store.dispatch({type: 'add', payload: 'test+2'});
         page.$store.dispatch({type: 'add', payload: 'test+3'});
         
         console.log(page.$store.getState().todo)
         expect(page.data.todoList.length).toBe(3);

         process.exit(1)

         page.$store.dispatch({type: 'del', payload: 'test+3'});
         expect(page.data.todoList.length).toBe(3);
         expect(app.$store.getState().todo.length).toBe(3);
         
         page.onHide();
         page.$store.dispatch({type: 'add', payload: 'test+2'});
         expect(page.data.todoList.length).toBe(3);
         expect(app.$store.getState().todo.length).toBe(4);
         
         page.onShow();
         page.onShow();
         expect(page.data.todoList.length).toBe(4);

         page.onUnload();
         page.onUnload();
    })

    test('page without map', ()=>{
        let page = {
            app,
            setData
        };

        reduxFnPage(page, 'Page');
        page.onLoad();
        page.onShow();
        page.$store.dispatch({type: 'add', payload: 'test+1'});
        expect(page.data).toBeFalsy();
    })

});

describe('Component', ()=>{
    let reduxFnApp = wxaRedux({
        reducers: combineReducers({todo})
    });
    let app = {}; reduxFnApp(app, 'App');
    
    let reduxFnComponent = wxaRedux({}, 'Component');

    global.getApp = ()=>app;

    test('mount redux', ()=>{
        const created = jest.fn();
        let com = {
            app,
            created
        };
        let comWithoutApp = {};
        let q = {scene: 110};

        reduxFnComponent(com, 'Component');
        expect(com.$store).not.toBeNull();
        com.created(q);
        expect(created.mock.calls.length).toBe(1);
        expect(created).toHaveBeenCalledWith(q);

        reduxFnComponent(comWithoutApp, 'Component');
        comWithoutApp.attached();
        expect(comWithoutApp.$unsubscribe).toBeFalsy();
    });

    test('dispatch data to component', ()=>{
        const attached = jest.fn();
        const detached = jest.fn();

        let com = {
            app,
            attached,
            detached,
            mapState: {
                todoList: (state)=>state.todo,
                first: (state)=>state.todo && state.todo[0]
            },
            setData
        }
        
        reduxFnComponent(com, 'Component');
        com.created();
        
        com.attached();
        expect(attached.mock.calls.length).toBe(1);
        expect(com.data).toBeFalsy();
        expect(com.$unsubscribe).not.toBeFalsy();

        com.$store.dispatch({type: 'add', payload: 'testcom'});
        expect(com.data.todoList.length).toBe(1);
        com.$store.dispatch({type: 'add', payload: 'testcom'});
        com.$store.dispatch({type: 'add', payload: 'testcom'});
        expect(com.data.todoList.length).toBe(3);
        expect(com.data.first).not.toBeFalsy();

        com.detached();
        expect(com.$unsubscribe).toBeNull();
    })
})

afterAll(()=>{
    global.getApp = og;
})