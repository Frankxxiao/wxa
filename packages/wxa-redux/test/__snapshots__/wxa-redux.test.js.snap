// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App register redux throw error while type undefined 1`] = `"不合法的wxa组件类型"`;
